### PR TITLE
fix random failing BasicAliasTest.testBug256837 #95

### PR DESCRIPTION
**Pull request by jukzi**

### CODE REVIEW

File: resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java [hunk 1]
```diff
@@ -410,6 +410,8 @@ public void testBug256837() throws CoreException {
 		// change the location of p2 project to the temp folder
 		replaceProject(p2, tempStore.toURI());
 
+		waitForRefresh(); // takes some milliseconds until aliasManager is notified
```
Comment (mickaelistria) [inline resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java:413]: What is the contract of AliasManager? Is it supposed to be async?
I'm afraid this test failure shows we actually have something broken in the synchronization.

Comment (jukzi) [reply resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java:413]: I don't know. If the AliasManager has any contract it is for sure broken multiple times: It does not even try to be thread safe and is NOT synchronized anywhere. There is totally no guarantee from JDK that it works.

what happens is that AliasManager.resourceChanged(IResourceChangeEvent) is called asynchronously and marks AliasManager.changedProjects = true . With a bit luck of having x86 memory architecture the main thread will see the change ;-)

This "fix" is the same as already  done for 
testBug232426, testBug358830, testCreateFolderInBackground, BasicAliasTest.testDeepLink, ...
which all suffer from the same problem.

I'd be happy if someone makes AliasManager thread safe + synchron.

**Comment by jukzi** [comment]
unrelated
```
 junit.framework.AssertionFailedError: 1.2: 
/testBuildDeltaUsingRelaxedRuleBug343256[~]: {} expected:<1> but was:<0>
org.eclipse.core.tests.internal.builders.RelaxedSchedRuleBuilderTest.testBuilderDeltaUsingRelaxedRuleBug343256(RelaxedSchedRuleBuilderTest.java:353)

```

File: resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java (hunk 1 above)
Comment (mickaelistria) [reply resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java:413]: > what happens is that AliasManager.resourceChanged(IResourceChangeEvent) is called asynchronously

That actually surprises me, I was kind of expecting POST_CHANGE resource listeners to happen synchronously. But maybe I build wrong assumption based on lucky side-effects all those years.
Looking at the code, I do not see where the resourceChanged() invocation becomes async. I see

Project.move() directly calls
Workspace.endOperation(), direcly calls
Workspace.broadcastPostChange(), directly calls
NotificationManager.notificationManager.broadcastChanges(tree, event, true), directly calls
NotificationManager.notify(), directly calls
IResourceChangeListener.resourceChanged(), and for AliasManager, directly changes the value of `changedProjects`, which is then used by AliasManager.computeAliases()

So can you please enlighten as to why this AliasManager would react asynchronously here? I'm missing something.

Comment (jukzi) [reply resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java:413]: I do not know WHY ... i am already happy that after much debugging i found that it does - if and only if the test fails  - which is was hard to reproduce (delay first ResourceException.provideStackTrace in main thread).

Something runs in parallel and its a race condition

Comment (laeubi) [reply resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java:413]: First thing I do is usually put `Thread.dumpStack()` into method where I' curious why and from where it is called

```
synchronized(Object.class) {
  Sysout(Thread.currentThread().getName());
  Thread.dumpStack();
}
```

Comment (jukzi) [reply resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java:413]: BTW you do not need to change the source for such but can put it as condition in a breakpoint.

**Comment by akurtakov** [comment]
@mickaelistria Can you help with making constantly failing test showing the bug?

**Comment by mickaelistria** [comment]
> Can you help with making constantly failing test showing the bug?

I don't have an idea how to achieve that without adding artificial test only configuration to AliasManager.
However, I think the test should work even with the auto-refresh stuff, the aliasManager most likely needs some synchronization guards.

**Comment by mickaelistria** [comment]
@jukzi I am spending more time on AliasManager, and I have some hope that if we do flag `changedProjects` as volatile, we might get help to reduce the problem here. What do you think? Would there be any risk or negative impact to check if we change that?

**Comment by jukzi** [comment]
typically volatile would be a good step in the right direction. However in this particular case it's not enough . the  "if (changedProjects){ ...changedProjects=false}" needs to be atomic operation and the datastructures must no be iterated while changed. 
- all those improvements however do not relate to the problem in this test.

**Comment by mickaelistria** [comment]
> all those improvements however do not relate to the problem in this test.

There is no problem in this test, there is a problem in AliasManager not being thread safe, and the test reveals the problem.

**Pull request closed without merging**